pkg/kubecost/allocation.go

```diff
@@ -266,34 +276,57 @@ func (parc ProportionalAssetResourceCost) Key(insertByNode bool) string {
 
 type ProportionalAssetResourceCosts map[string]ProportionalAssetResourceCost
 
-func (parcs ProportionalAssetResourceCosts) Insert(parc ProportionalAssetResourceCost, insertByNode bool) {
+func (parcs ProportionalAssetResourceCosts) Clone() ProportionalAssetResourceCosts {
+	cloned := ProportionalAssetResourceCosts{}
+
+	for key, parc := range parcs {
+		cloned[key] = parc.Clone()
+	}
+	return cloned
+}
+
+func (parcs ProportionalAssetResourceCosts) Insert(parc ProportionalAssetResourceCost, insertByNode, isAccumulation bool) {
 	if !insertByNode {
 		parc.Node = ""
 		parc.ProviderID = ""
 	}
 	if curr, ok := parcs[parc.Key(insertByNode)]; ok {
+
+		CPUPercentage := curr.CPUPercentage + parc.CPUPercentage
```

Comment:
Can you explain the logic/math behind adding/averaging the percentages?

Reply:
@nik-kc does my comment @ https://github.com/opencost/opencost/pull/1867/files#diff-7c339daaa6b130f0e74be88b76e4c35d3941edf78f886fae793772b8f566c58cR298 make sense? happy to add further detail if not

Reply:
I think what's confusing me is why adding percentages works to begin with, and by extension, why averaging them would also work, given that the windows could be different. I'm having a hard time seeing how that math works out.

Reply:
I see. So basically, there are two different dimensions we are summing across: resource and time. If pod A and pod B at the same instant are using 1% each of a node's CPU, then when we combine their allocations, it makes sense to sum them for a total of 2%. That's the first dimension and what this highlighted line handles. 

The second dimension is time. If pod A+B is using 2% each min, when we accumulate those (which is just a time based aggregation under the hood), over 10 minutes then if we just sum, it looks like Pod A+B is using 20% of the node... when in reality it was just using 2% each minute over the 10 minute period. In the time case, we want the _average_ percentage that it used over the period. 

Because of this, we have no choice but to pass down the context to the addition function, and tell it whether we want the average (when we are combining these over time) or the sum (when we are combining these across resources)

Reply:
That makes sense! Since I'm not super familiar with what the data is expected to look like, would it be valid in the non-accumulation case to have a percentage over 100%? e.g. given pod A and B in your first example, if they had percentages of 80% and 60%, respectively.

Reply:
my understanding is that would be an acceptable outcome due to exactly that reason - at different times, pods can consume a majority of resources, so if those pods get aggregated together, the usage would be > 100%

Reply:
Awesome, makes sense. Thanks for explaining!

Reply:
I think I haven't followed this fully. Let me pose an example and see how it works out. I'll stick with just CPU to avoid overloading with identical math.

Say I have a **one-node cluster** with **three Pods** (each with one container). The **node has 4 CPU**.

Over the span of hour 1:
- Pod 1 runs for 30 minutes, consuming 1 CPU while alive. PARC: 12.5% (0.5 core-hours / 4 available core-hours)
- Pod 2 runs for 1 hour, consuming 2 CPU while alive. PARC: 50% (2 core-hours)
- Pod 3 runs for 1 hour, consuming 1 CPU while alive. PARC: 25% (1 core-hour)

Over the span of hour 2:
- Pod 1 does not run. PARC: 0% (0 core-hours / 4 available core-hours)
- Pod 2 runs for 30 minutes, consuming 2 CPU while active. PARC: 25% (1 core-hour)
- Pod 3 runs for 1 hour, consuming 1 CPU while active. PARC: 25% (1 core-hour)

Over the span of hour 3:
- Pod 1 does not run. PARC: 0% (0 core-hours / 4 available)
- Pod 2 runs for 30 minutes, consuming 3 CPU while active. PARC: 37.5% (1.5 core-hours)
- Pod 3 runs for 1 hour, consuming 1 CPU while active. PARC: 25% (1 core-hour)

I'd expect the following accumulated PARC:
- Pod 1: (0.5 + 0 + 0) core-hours used / (4 + 4 + 4) core-hours available = 0.5/12 = **4.16%**
- Pod 2: (2 + 1 + 1.5) / (4 + 4 + 4) = 4.5/12 = **37.5%**
- Pod 3: (1 + 1 + 1) / (4 + 4 + 4) = 3/12 = **25%**

The logic I'm seeing in this PR will do this:
- Pod 1: `12.5%`-> `(12.5% + 0%) / 2` -> `((12.5% + 0%) / 2) + 0%) / 2` = **3.125%**
- Pod 2: `50%` -> `(50% + 37.5%) / 2` -> `((50% + 37.5%) / 2) + 37.5%) / 2` = **40.625%**
- Pod 3: `25%` -> `(25% + 25%) / 2` -> `((25% + 25%) / 2) + 25%) / 2` = **25%**

This leads me to think that the logic in this PR is incorrect. However, I may be misunderstanding what PARC is supposed to represent.

Reply:
Before I forget: what if the node (and workloads) go away halfway through hour 3? How does that affect PARC by node? And by cluster?

Reply:
This has been remediated by tracking the individual components comprising the PARC during the computation process. We have a unit test covering this exact scenario and the numbers match now